Gmail notifications support through new imcb_notify_email() API

dequis opened the pull request:
Rebased and reindented patch from http://bugs.bitlbee.org/bitlbee/ticket/647 by `Artem Savkov <artem.savkov@gmail.com>` aka trac3r aka @sm00th

Not tested or changed other than the rebasing, just pullrequesting it for review here. The commit message could use some improvement, it's just the first thing i came up with.

dequis commented on doc/user-guide/commands.xml:
```diff
@@ -1021,6 +1021,28 @@
+
+		<description>
+			<para>
+				Same as "mail_notofications" but GMail specific.
```
lowest hanging fruit ever: a typo

dequis commented on protocols/jabber/iq.c:
```diff
@@ -723,6 +776,62 @@ xt_status jabber_iq_query_server(struct im_connection *ic, char *jid, char *xmln
+		return XT_HANDLED;
+	}
+	if (strcmp(xmlns, XMLNS_GMAILNOTIFY) == 0) {
+		unsigned long long l_time = 0;
```
`guint64`

dequis commented on protocols/jabber/iq.c:
```diff
@@ -723,6 +776,62 @@ xt_status jabber_iq_query_server(struct im_connection *ic, char *jid, char *xmln
+			}
+
+			thread = c->children;
+			subject = xt_find_node(thread, "subject")->text;
```
this looks pretty and it would be neat to be able to do this safely, but it could dereference a null there if there's no subject

dequis commented on protocols/jabber/iq.c:
```diff
@@ -723,6 +776,62 @@ xt_status jabber_iq_query_server(struct im_connection *ic, char *jid, char *xmln
+		if (l_time && (!jd->gmail_time || l_time > jd->gmail_time)) {
+			jd->gmail_time = l_time;
+			if (tid) {
+				jd->gmail_tid = g_strdup(tid);
```
this leaks on logout and if there's an existing value of jd->gmail_tid

dequis commented on protocols/jabber/jabber.h:
```diff
@@ -100,6 +101,8 @@ struct jabber_data {
 	   presence_send_update() to inform the server about the changes. */
 	const struct jabber_away_state *away_state;
 	char *away_message;
+	unsigned long long gmail_time;
```
guint64 here too

dequis commented on protocols/yahoo/yahoo.c:
```diff
@@ -948,13 +958,19 @@ void ext_yahoo_game_notify(int id, const char *me, const char *who, int stat, co
 	struct im_connection *ic = byahoo_get_ic_by_id(id);
+	char *msg;
 
 	if (!set_getbool(&ic->acc->set, "mail_notifications")) {
```
could invert this condition

dequis commented on protocols/jabber/iq.c:
```diff
@@ -702,6 +729,32 @@ xt_status jabber_iq_parse_features(struct im_connection *ic, struct xt_node *nod
+	node = xt_new_node("query", NULL, NULL);
+	xt_add_attr(node, "xmlns", XMLNS_GMAILNOTIFY);
+	if (jd->gmail_time) {
+		xt_add_attr(node, "newer-than-time", g_strdup_printf("%llu", (jd->gmail_time + 1)));
```
g_strdup_printf leaks here

dequis commented on protocols/jabber/iq.c:
```diff
@@ -702,6 +729,32 @@ xt_status jabber_iq_parse_features(struct im_connection *ic, struct xt_node *nod
+		xt_add_attr(node, "newer-than-tid", jd->gmail_tid);
+	}
+
+	if (!(query = jabber_make_packet("iq", "get", g_strdup_printf("%s@%s", jd->username, jd->server), node))) {
```
g_strdup_printf also leaks here

dequis commented on protocols/jabber/iq.c:
```diff
@@ -723,6 +776,62 @@ xt_status jabber_iq_query_server(struct im_connection *ic, char *jid, char *xmln
+			struct xt_node *thread;
+			unsigned long long t_time;
+
+			t_time = strtoull(xt_find_attr(c, "date"), NULL, 10);
```
glib provides `g_ascii_strtoull`, which isn't locale-sensitive.

jgeboski commented:
What is this waiting on?

dequis commented:
Mostly testing/valgrinding etc, but extra review would be appreciated. Also just noticed the merge conflict (not surprising, given how it touches msn stuff)

TBH I kinda forgot about this one, but it's been waiting for years in trac so a few weeks here can't hurt :3

jgeboski commented on protocols/jabber/jabber.c:
```diff
@@ -331,8 +342,10 @@ static void jabber_logout(struct im_connection *ic)
 	md5_free(&jd->cached_id_prefix);
 
 	g_free(jd->oauth2_access_token);
+	g_free(jd->oauth2_access_token);
```
Double free.

dequis replied:
Double free as in double freedom!

(how the hell did that happen? must be some lazy conflict resolution)

dequis commented:
This needs some limit.

![](http://dump.dequis.org/WZb8Y.png)

(there's half a screen more afterwards)

jgeboski commented:
Might be cleaner to get have a "GMail" user send the messages.  This way at least the user can redirect the output somewhere.

dequis commented:
That's configurable already

>notify_handle
>This setting is available for protocols with e-mail notification functionality. If set to empty all e-mail notifications will go to control channel, if set to some string - this will be the name of a contact who will PRIVMSG you on every new notification

dequis commented:
Made some commits here. Added a `gmail_notifications_limit` setting (need opinion on this), removed the "snippet" after emails, made it look up the sender in the xml instead of repeating your own email every time, etc

Before:

```
00:16 <@root> jabber - Logging in: New mail for me@irl. Subj: New private message has arrived
00:16 <@root> jabber - Logging in: Hello dequis, You have received a new private message from "awaps" to your account on
00:16 <@root> jabber - Logging in: New mail for me@irl. Subj: Learn Now: 100 Courses On Demand
00:16 <@root> jabber - Logging in: coursera Learning Without Limits Discover On Demand Over 100 of Coursera's top courses are now
00:16 <@root> jabber - Logging in: New mail for me@irl. Subj: Celebrate .CLUB's 1 Day $1 Sale!
00:16 <@root> jabber - Logging in: .CLUB Anniversary 1 Day $1 Sale Enjoy $1 .CLUB registrations in honor of their 1 year anniversary! No
00:16 <@root> jabber - Logging in: New mail for me@irl. Subj: 8 games from Surprise Attack in the Humble Weekly Bundle
00:16 <@root> jabber - Logging in: GAH don't sneak up on us like that. We, phew, thought you were like, hah, a monster or something.
00:16 <@root> jabber - Logging in: New mail for me@irl. Subj: [qtile-dev] Use greek letters as group name?
00:16 <@root> jabber - Logging in: Qtile loads the default config whenever I try to set group names to λ,Ω,Δ (or any greek alphabet) Any
00:16 <@root> jabber - Logging in: New mail for me@irl. Subj: Android M podría presentarse en el evento Google I/O 2015 - El Sony Xperia Z4 Compact se podría revelar el 13 de mayo - Review: Huawei Y600 y TalkBand - ¡y más!
00:16 <@root> jabber - Logging in: 7 de mayo de 2015 ¿Problemas con la visualización? Clic aquí Android M podría presentarse en el
```

After:

```
00:44 <@root> jabber - Logging in: New mail from Coursera: Learn Now: 100 Courses On Demand
00:44 <@root> jabber - Logging in: New mail from Dynadot: Celebrate .CLUB's 1 Day $1 Sale!
00:44 <@root> jabber - Logging in: New mail from Humble Bundle: 8 games from Surprise Attack in the Humble Weekly Bundle
00:44 <@root> jabber - Logging in: New mail from Logan Kaser: [qtile-dev] Use greek letters as group name?
00:44 <@root> jabber - Logging in: New mail from RedUSERS Mobile: Android M podría presentarse en el evento Google I/O 2015 - El Sony Xperia Z4 Compact se podría revelar el 13 de mayo - Review: Huawei Y600 y TalkBand - ¡y más!
00:44 <@root> jabber - Logging in: Logged in
```

I had a counter there for the remaining emails, but it turns out the server limits it at 30, so it was pointless to say "25 more" every time.

**Also, in 3d2004629be5d818289703982fb6096acb7462b1, I changed the declaration of `imcb_notify_email` completely**, now it looks more like imcb_log. Saves some messing with `g_strdup_printf` for the callers, and `flags`/`sent_at` weren't used anyway.

REVIEWINGS REQUESTED

jelmer commented:
You get a lot of interesting mail.

dequis commented:
Thanks! I am proud of my mail receiving choices - even if i could receive different mails, i wouldn't choose anything else.

*(this feature would be more useful if it only showed the 'primary' tab of gmail, excluding 'promotions', but i don't think i want to waste more time with this PR)*

dequis commented:
REVIEWINGS PENDING

And, like, review of the code, not of my emails. I can review those myself.

Wilm0r commented on doc/user-guide/commands.xml:
```diff
@@ -1022,6 +1022,39 @@
+
+	</bitlbee-setting>
+
+	<bitlbee-setting name="gmail_notifications" type="boolean" scope="account">
```
Why add a GMail-specific one?

Something tells me I will find out later on ... Because this seems sub-optimal to me.

Wilm0r commented:
I should've looked and responded earlier, sorry. :-(

I'd like this to be more uniform. Accounts that support mail notifications have a setting called mail_notifications, accounts that don't, don't.

Then imcb_blah() suppresses any notifications if that setting is disabled. And at login time if some msg to the srv is required to get these notifications shown, read the setting so you know whether to send it or not.

Is there a reason this one can't work that way?

dequis commented:
I think jabber is the main problem (with the setting naming). The gmail stuff is a google specific extension that isn't intended to become standard ever, so enabling that in a non-google jabber account would be a no-op. I guess that's not so bad...

>Then imcb_blah() suppresses any notifications if that setting is disabled.

So, move the setting check to imcb_notify_email()?

dequis commented:
(I just noticed I opened this PR two months ago. Huh)

Anyway I noticed that the gmail stuff actually checks for server capabilities before enabling itself, so it definitely would be a no-op for non-gmail servers, but I sorta expected it to send garbage to the server (it doesn't). So i just renamed the settings to the generic name

So now `imcb_notify_email` has one check for the `mail_notifications` setting

I should squash this mess somehow.
